File: dp/CoinChange.java
/*
 * Find minimum number of coins
 * https://leetcode.com/problems/coin-change/
 */

public class CoinChange {

    /**
     * @ return minimum numbre of coins for the sum
     */
    private static int[] coins = {1, 2, 5};

    public static int countCoinChange(int target ) {

        if (target <= 0) {
            return 0;
        }
        int smallest = Integer.MAX_VALUE;

        for (int i = 0; i < coins.length; i++) {
         smallest = Math.min(smallest, countCoinChange(target - coins[i]) + 1);
        }

        return smallest;
    }

    public static void main(String[] args) {
        assert countCoinChange(11) == 3;

        assert countCoinChange(0) == 0;

        assert countCoinChange(1) == 1;

    }
}

File: leetcode/coin_change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        
        max = amount + 1 
        dp = [max] * (amount + 1)
        dp[0] = 0
        
        for i in range(1 , amount + 1):     
            for coin in coins:
                if coin <= amount:
                    dp[i] = min(dp[i], 1 + dp[i - coin])
                    
                    
        return -1 if dp[amount] > amount else dp[amount]
